### PR TITLE
Added RawPacketSendEvent, RawPacketReceiveEvent, and updates for Server,

### PR DESCRIPTION
**Pull request by LuciaPE**
### Description

Created raw packet send, receive Event.

### Reason to modify

By this PR, plugins will be able to handle raw packets, also it can be used on server-server connection and so on.

### Tests & Reviews

It has been tested on my computer.

### CODE REVIEW

File: src/pocketmine/network/Network.php [hunk 1]
```diff
@@ -286,6 +287,9 @@ public function getPacket($id) {
 	 * @param string $payload
 	 */
 	public function sendPacket($address, $port, $payload) {
+		$ev = new RawPacketSendEvent($address, $port, $payload);
+		$this->server->getPluginManager()->callEvent($ev);
+		if(! $ev->isCancelled())
```
Comment (ishitatsuyuki) [inline src/pocketmine/network/Network.php]: Formatting issue

File: src/pocketmine/Server.php [hunk 2]
```diff
@@ -168,6 +168,7 @@
 use pocketmine\utils\Utils;
 use pocketmine\utils\UUID;
 use pocketmine\utils\VersionString;
+use pocketmine\event\server\RawPacketReceiveEvent;
```
Comment (ishitatsuyuki) [inline src/pocketmine/Server.php]: Keep it sorted

File: src/pocketmine/Server.php [hunk 3]
```diff
@@ -2686,6 +2687,9 @@ private function titleTick(){
 	public function handlePacket($address, $port, $payload){
 		try{
 			if(strlen($payload) > 2 and substr($payload, 0, 2) === "\xfe\xfd" and $this->queryHandler instanceof QueryHandler){
+				$ev = new RawPacketReceiveEvent($address, $port, $payload);
+				$this->getPluginManager()->callEvent($ev);
+				if(!$ev->isCancelled())
```
Comment (ishitatsuyuki) [inline src/pocketmine/Server.php]: Dumb formatting

File: src/pocketmine/network/Network.php [hunk 4]
```diff
@@ -97,6 +97,7 @@
 use pocketmine\utils\Binary;
 use pocketmine\utils\BinaryStream;
 use pocketmine\utils\MainLogger;
+use pocketmine\event\server\RawPacketSendEvent;
```
Comment (ishitatsuyuki) [inline src/pocketmine/network/Network.php]: Keep it sorted

**Comment by JackNoordhuis** [comment]
I see no use for these events; all they will do is put more load on the server.

File: src/pocketmine/network/Network.php (hunk 4 above)
Comment (LuciaPE) [inline src/pocketmine/network/Network.php]: Got it

**Comment by LuciaPE** [comment]
@JackNoordhuis It can be used on connecting server to server by plugin.

**Comment by JackNoordhuis** [comment]
@LuciaPE You can use [DataPacketReceiveEvent](https://github.com/iTXTech/Genisys/blob/master/src/pocketmine/event/server/DataPacketReceiveEvent.php) and [DataPacketSendEvent](https://github.com/iTXTech/Genisys/blob/master/src/pocketmine/event/server/DataPacketSendEvent.php)

**Comment by jojoe77777** [comment]
@JackNoordhuis That's for player packets, they want server communication via plugins

File: src/pocketmine/event/server/RawPacketReceiveEvent.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
Comment (dktapps) [inline src/pocketmine/event/server/RawPacketReceiveEvent.php]: Missing licence header

File: src/pocketmine/event/server/RawPacketReceiveEvent.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+use pocketmine\event;
+use pocketmine\event\Cancellable;
+
+class RawPacketReceiveEvent extends ServerEvent implements Cancellable {
+	public static $handlerList = null;
+	private $payload;
+	private $address;
+	private $port;
+	public function __construct(string $address, int $port, string $payload) {
```
Comment (dktapps) [inline src/pocketmine/event/server/RawPacketReceiveEvent.php]: where is the whitespace? this is all squashed together, it's unpleasant to look at.

**Comment by JackNoordhuis** [comment]
@jojoe77777 The only packets that are sent through Genisys itself are to clients. Synapse has been completely stripped and moved to its own plugin and the server list query is handled through the [QueryRegenerateEvent](https://github.com/iTXTech/Genisys/blob/master/src/pocketmine/event/server/QueryRegenerateEvent.php).

**Comment by jojoe77777** [comment]
@JackNoordhuis I know, but they want to implement plugin messaging...

**Comment by LinuxLogo** [comment]
@JackNoordhuis Jesus kid.. you really are a prick...

File: src/pocketmine/Server.php [hunk 7]
```diff
@@ -2686,6 +2687,9 @@ private function titleTick(){
 	public function handlePacket($address, $port, $payload){
 		try{
 			if(strlen($payload) > 2 and substr($payload, 0, 2) === "\xfe\xfd" and $this->queryHandler instanceof QueryHandler){
+				$ev = new RawPacketReceiveEvent($address, $port, $payload);
```
Comment (dktapps) [inline src/pocketmine/Server.php]: The way you have added this here will only allow raw packet handling by plugins if it's a Query packet if I am reading this correctly. That doesn't seem particularly useful.

Comment (LinuxLogo) [inline src/pocketmine/Server.php]: Hmmm that I didn't notice.. but you should add snapse back into Genisys and/or some sort of Nemisys Impementation. Please and thanks.

Comment (LinuxLogo) [inline src/pocketmine/Server.php]: Oi, Synapse*

**Pull request closed without merging**